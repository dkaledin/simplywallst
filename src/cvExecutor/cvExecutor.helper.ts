import {CvExecutorStep, IResultByCurrentStep} from './cvExecutor.interface';

export function getResultByCurrentStep(currentStep: CvExecutorStep, command: string): IResultByCurrentStep {
    return steps[currentStep](command);
}

export function getNextByStep(step: CvExecutorStep) {
    const promptByStep = {
        [CvExecutorStep.JobTitle]: 'job_title: (Front-end Developer)',
        [CvExecutorStep.GivenName]: 'given_name:',
        [CvExecutorStep.FamilyName]: 'family_name:',
        [CvExecutorStep.Email]: 'email:',
        [CvExecutorStep.Mobile]: 'mobile:',
        [CvExecutorStep.PortfolioUrl]: 'portfolio_url:',
        [CvExecutorStep.Bio]: 'bio:',
        [CvExecutorStep.Skills]: 'skills:',
        [CvExecutorStep.SkillsYears]: 'years:',
        [CvExecutorStep.SkillsConfidenceRating]: 'confidence_rating:',
        [CvExecutorStep.SkillsTechnology]: 'technology:',
        [CvExecutorStep.SkillsConfirm]: 'Do You have any more skills? (yes)',
        [CvExecutorStep.PreviousJobs]: 'previous_jobs:',
        [CvExecutorStep.PreviousJobsTitle]: 'job_title:',
        [CvExecutorStep.PreviousJobsYears]: 'years:',
        [CvExecutorStep.PreviousJobsConfirm]: 'Do You have any more previous jobs? (yes)',
        [CvExecutorStep.JsonConfirm]: 'Is this Ok? (yes)',
        [CvExecutorStep.SendingJson]: '',
        [CvExecutorStep.Aborted]: '',
    };

    return {
        step,
        prompt: promptByStep[step],
    };
}

export const steps = {
    [CvExecutorStep.Init]: getInit,
    [CvExecutorStep.JobTitle]: getJobTitle,
    [CvExecutorStep.GivenName]: getGivenName,
    [CvExecutorStep.FamilyName]: getFamilyName,
    [CvExecutorStep.Email]: getEmail,
    [CvExecutorStep.Mobile]: getMobile,
    [CvExecutorStep.PortfolioUrl]: getPortfolioUrl,
    [CvExecutorStep.Bio]: getBio,
    [CvExecutorStep.Skills]: getSkills,
    [CvExecutorStep.SkillsYears]: getSkillsYears,
    [CvExecutorStep.SkillsConfidenceRating]: getSkillsConfidenceRating,
    [CvExecutorStep.SkillsTechnology]: getSkillsTechnology,
    [CvExecutorStep.SkillsConfirm]: getSkillsConfirm,
    [CvExecutorStep.PreviousJobs]: getPreviousJobs,
    [CvExecutorStep.PreviousJobsTitle]: getPreviousJobsTitle,
    [CvExecutorStep.PreviousJobsYears]: getPreviousJobsYears,
    [CvExecutorStep.PreviousJobsConfirm]: getPreviousJobsConfirm,
    [CvExecutorStep.JsonConfirm]: getJsonConfirm,
    [CvExecutorStep.SendingJson]: getSendingJson,
    [CvExecutorStep.Aborted]: getAborted,
};

export function toFormattedJson(object: any): string {
    return JSON.stringify(object, undefined, 2);
}

function getInit() {
    return {
        message: ['Welcome to the CV terminal command!'],
        next: getNextByStep(CvExecutorStep.JobTitle),
    };
}

function getJobTitle(command: string) {
    return {
        result: {
            job_title: command || 'Front-end Developer',
        },
        next: getNextByStep(CvExecutorStep.GivenName),
    };
}

function getGivenName(command: string) {
    if (command !== '') {
        return {
            result: {
                given_name: command,
            },
            next: getNextByStep(CvExecutorStep.FamilyName),
        };
    } else {
        return {
            error: 'Wrong given_name',
            next: getNextByStep(CvExecutorStep.GivenName),
        };
    }
}

function getFamilyName(command: string) {
    if (command !== '') {
        return {
            result: {
                family_name: command,
            },
            next: getNextByStep(CvExecutorStep.Email),
        };
    } else {
        return {
            error: 'Wrong family_name',
            next: getNextByStep(CvExecutorStep.FamilyName),
        };
    }
}

function getEmail(command: string) {
    if (isValidEmail(command)) {
        return {
            result: {
                email: command,
            },
            next: getNextByStep(CvExecutorStep.Mobile),
        };
    } else {
        return {
            error: 'Wrong email',
            next: getNextByStep(CvExecutorStep.Email),
        };
    }
}

function getMobile(command: string) {
    if (/^[0-9 ]+$/i.test(command)) {
        return {
            result: {
                mobile: command,
            },
            next: getNextByStep(CvExecutorStep.PortfolioUrl),
        };
    } else {
        return {
            error: 'Wrong mobile',
            next: getNextByStep(CvExecutorStep.Mobile),
        };
    }
}

function getPortfolioUrl(command: string) {
    if (isValidUrl(command)) {
        return {
            result: {
                portfolio_url: command,
            },
            next: getNextByStep(CvExecutorStep.Bio),
        };
    } else {
        return {
            error: 'Wrong portfolio_url',
            next: getNextByStep(CvExecutorStep.PortfolioUrl),
        };
    }
}

function getBio(command: string) {
    if (command !== '') {
        return {
            result: {
                bio: command,
            },
            next: getNextByStep(CvExecutorStep.Skills),
        };
    } else {
        return {
            error: 'Wrong portfolio_url',
            next: getNextByStep(CvExecutorStep.Bio),
        };
    }
}

let skillKey = '';

function getSkills(command: string) {
    if (command !== '') {
        skillKey = command;
        return {
            result: {
                skills: {
                    [command]: {},
                },
            },
            next: getNextByStep(CvExecutorStep.SkillsYears),
        };
    } else {
        return {
            error: 'Wrong skill name',
            next: getNextByStep(CvExecutorStep.Skills),
        };
    }
}

function getSkillsYears(command: string) {
    if (/^[0-9]+$/.test(command)) {
        return {
            result: {
                skills: {
                    [skillKey]: {
                        years: parseInt(command, 10),
                    },
                },
            },
            next: getNextByStep(CvExecutorStep.SkillsConfidenceRating),
        };
    } else {
        return {
            error: 'Please enter valid skills years',
            next: getNextByStep(CvExecutorStep.SkillsYears),
        };
    }
}

function getSkillsConfidenceRating(command: string) {
    if (/^[0-9]{1,2}$/.test(command)) {
        return {
            result: {
                skills: {
                    [skillKey]: {
                        confidence_rating: parseInt(command, 10),
                    },
                },
            },
            next: getNextByStep(CvExecutorStep.SkillsTechnology),
        };
    }
    return {
        error: 'Please enter valid skills confidence_rating',

        next: getNextByStep(CvExecutorStep.SkillsConfidenceRating),
    };
}

function getSkillsTechnology(command: string) {
    if (command !== '') {
        return {
            result: {
                skills: {
                    [skillKey]: {
                        technology: command.split(/\s?,\s?/),
                    },
                },
            },
            next: getNextByStep(CvExecutorStep.SkillsConfirm),
        };
    }

    return {
        error: 'Please enter valid skills technology',
        next: getNextByStep(CvExecutorStep.SkillsTechnology),
    };
}

function getSkillsConfirm(command: string) {
    if (command === '' || /yes|ok|true/i.test(command)) {
        return {
            next: getNextByStep(CvExecutorStep.Skills),
        };
    }

    return {
        message: ['Great!'],
        next: getNextByStep(CvExecutorStep.PreviousJobs),
    };
}

let previousJobsKey = '';

function getPreviousJobs(command: string) {
    if (command !== '') {
        previousJobsKey = command;
        return {
            result: {
                previous_jobs: {
                    [command]: {},
                },
            },
            next: getNextByStep(CvExecutorStep.PreviousJobsTitle),
        };
    }
    return {
        error: 'Please enter valid previous_jobs',
        next: getNextByStep(CvExecutorStep.PreviousJobs),
    };
}

function getPreviousJobsTitle(command: string) {
    if (command !== '') {
        return {
            result: {
                previous_jobs: {
                    [previousJobsKey]: {
                        job_title: command,
                    },
                },
            },
            next: getNextByStep(CvExecutorStep.PreviousJobsYears),
        };
    }
    return {
        error: 'Please enter valid job_title',
        next: getNextByStep(CvExecutorStep.PreviousJobsTitle),
    };

}

function getPreviousJobsYears(command: string) {
    if (/^[0-9]+$/.test(command)) {
        return {
            result: {
                previous_jobs: {
                    [previousJobsKey]: {
                        years: parseInt(command, 10),
                    },
                },
            },
            next: getNextByStep(CvExecutorStep.PreviousJobsConfirm),
        };
    }
    return {
        error: 'Please enter valid year',
        next: getNextByStep(CvExecutorStep.PreviousJobsYears),
    };
}

function getPreviousJobsConfirm(command: string) {
    if (command === '' || /yes|ok|true/i.test(command)) {
        return {
            next: getNextByStep(CvExecutorStep.PreviousJobs),
        };
    }

    return {
        message: ['Json is ready to send!', 'Please check it.'],
        next: getNextByStep(CvExecutorStep.JsonConfirm),
    };
}

function getJsonConfirm(command: string) {
    if (command === '' || /yes|ok|true/i.test(command)) {
        return {
            next: getNextByStep(CvExecutorStep.SendJson),
        };
    }
    return {
        next: getNextByStep(CvExecutorStep.Aborted),
    };
}

function getSendingJson() {
    return {
        next: getNextByStep(CvExecutorStep.SendingJson),
    };
}

function getAborted() {
    return {
        next: getNextByStep(CvExecutorStep.Aborted),
    };
}

function isValidEmail(email: string): boolean {
    // tslint:disable-next-line
    const pattern = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

    return pattern.test(email);
}

function isValidUrl(url: string): boolean {
    // tslint:disable-next-line
    const pattern = /((([A-Za-z]{3,9}:(?:\/\/)?)(?:[\-;:&=\+\$,\w]+@)?[A-Za-z0-9\.\-]+|(?:www\.|[\-;:&=\+\$,\w]+@)[A-Za-z0-9\.\-]+)((?:\/[\+~%\/\.\w\-_]*)?\??(?:[\-\+=&;%@\.\w_]*)#?(?:[\.\!\/\\\w]*))?)/;

    return pattern.test(url);
}
