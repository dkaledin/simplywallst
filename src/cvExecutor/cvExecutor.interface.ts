export interface IResultByCurrentStep {
    result?: {};
    message?: [];
    error?: string;
    next: {
        step: CvExecutorStep,
        prompt: string,
    };
}

export enum CvExecutorStep {
    Init,
    JobTitle,
    GivenName,
    FamilyName,
    Email,
    Mobile,
    PortfolioUrl,
    Bio,
    Skills,
    SkillsYears,
    SkillsConfidenceRating,
    SkillsTechnology,
    SkillsConfirm,
    PreviousJobs,
    PreviousJobsTitle,
    PreviousJobsYears,
    PreviousJobsConfirm,
    JsonConfirm,
    SendJson,
    SendingJson,
    Aborted,
}

export interface ISuccessJsonResponse {
    success: string;
}

export interface IErrorsJsonResponse {
    errors: IErrorJsonResponse[];
}

interface IErrorJsonResponse {
    status: number;
    source: {
        parameter: string;
    };
    title: string;
    detail: string;
}
