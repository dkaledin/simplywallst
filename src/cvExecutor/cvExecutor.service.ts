export function sendJsonCv(json: string): Promise<Response> {
    return fetch('https://api.simplywall.st/api/careers/job/apply', {
        method: 'POST',
        cache: 'no-cache',
        headers: {
            'Content-Type': 'application/json',
        },
        body: json,
    });
}
