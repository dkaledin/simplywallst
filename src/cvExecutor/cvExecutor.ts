import {BaseExecutor, SetState} from 'react-simply-terminal-emulator';
import merge from 'lodash.merge';
import {getResultByCurrentStep, toFormattedJson} from './cvExecutor.helper';
import {CvExecutorStep, IErrorsJsonResponse, ISuccessJsonResponse} from './cvExecutor.interface';
import {sendJsonCv} from './cvExecutor.service';

export class CvExecutor extends BaseExecutor {
    private setState: SetState;
    private prevStep: CvExecutorStep;
    private complete: boolean;
    private result = {};

    public init() {
        this.prevStep = CvExecutorStep.Init;
        this.complete = false;
    }

    public execute(command: string, setState: SetState): void {
        this.setState = setState;

        const currentStepData = getResultByCurrentStep(this.prevStep, command);
        const {
            next,
            error,
        } = currentStepData;

        const message = currentStepData.message || [];
        const result = currentStepData.result || {};

        if (next.step === CvExecutorStep.JsonConfirm) {
            this.updateState([toFormattedJson(this.result)], next.prompt);
            this.prevStep = next.step;

        } else if (next.step === CvExecutorStep.SendJson) {
            this.updateState(['Sending...'], '');
            this.prevStep = CvExecutorStep.SendingJson;

            sendJsonCv(JSON.stringify(this.result)).then(response => {
                this.complete = true;
                const responseMessage: string[] = [];

                response.json().then(json => {
                    if (response.status === 200) {
                        responseMessage.push((json as ISuccessJsonResponse).success);
                    } else {
                        (json as IErrorsJsonResponse).errors
                            .forEach(responseError => {
                                responseMessage.push(`${responseError.title}: ${responseError.detail}`);
                            });
                    }
                    this.updateState(responseMessage);
                });
            });

        } else if (next.step === CvExecutorStep.Aborted) {
            this.complete = true;
            this.updateState(['Aborted']);

        } else {
            if (error === undefined) {
                this.updateState(message, next.prompt);
                this.prevStep = next.step;
                this.result = merge(this.result, result);

            } else {
                this.updateState([error], next.prompt);
            }
        }
    }

    public isCompleted(): boolean {
        return this.complete;
    }

    protected getCommandPattern() {
        return /^cv$/i;
    }

    private updateState(newLines: string[], prompt?: string) {
        this.setState((prevState, props) => {
            const nextPrompt = prompt === undefined
                ? props.prompt
                : prompt;

            return {
                history: [
                    ...prevState.history,
                    ...newLines,
                ],
                prompt: nextPrompt,
            };
        });
    }
}
