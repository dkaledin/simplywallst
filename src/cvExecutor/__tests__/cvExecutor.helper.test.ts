import {getResultByCurrentStep} from '../cvExecutor.helper';
import {CvExecutorStep} from '../cvExecutor.interface';

describe('getResultByCurrentStep', () => {
    it('should return expected response for every receive command', () => {
        const actualList = [];

        actualList.push(getResultByCurrentStep(CvExecutorStep.Init, ''));
        actualList.push(getResultByCurrentStep(CvExecutorStep.JobTitle, ''));
        actualList.push(getResultByCurrentStep(CvExecutorStep.GivenName, 'Warren'));
        actualList.push(getResultByCurrentStep(CvExecutorStep.FamilyName, 'Buffett'));
        actualList.push(getResultByCurrentStep(CvExecutorStep.Email, 'tupaclives1985@gmail.com'));
        actualList.push(getResultByCurrentStep(CvExecutorStep.Mobile, '04 0404 4044'));
        actualList.push(getResultByCurrentStep(CvExecutorStep.PortfolioUrl, 'https://warrenbuffett.id.au'));
        actualList.push(getResultByCurrentStep(CvExecutorStep.Bio, 'Hi I like building cool things...'));

        actualList.push(getResultByCurrentStep(CvExecutorStep.Skills, 'php'));
        actualList.push(getResultByCurrentStep(CvExecutorStep.SkillsYears, '2'));
        actualList.push(getResultByCurrentStep(CvExecutorStep.SkillsConfidenceRating, '50'));
        actualList.push(getResultByCurrentStep(CvExecutorStep.SkillsTechnology, 'symfony, laravel'));
        actualList.push(getResultByCurrentStep(CvExecutorStep.SkillsConfirm, 'yes'));

        actualList.push(getResultByCurrentStep(CvExecutorStep.Skills, 'js'));
        actualList.push(getResultByCurrentStep(CvExecutorStep.SkillsYears, '3'));
        actualList.push(getResultByCurrentStep(CvExecutorStep.SkillsConfidenceRating, '99'));
        actualList.push(getResultByCurrentStep(CvExecutorStep.SkillsTechnology, 'typescript,react , redux, node'));
        actualList.push(getResultByCurrentStep(CvExecutorStep.SkillsConfirm, 'no'));

        actualList.push(getResultByCurrentStep(CvExecutorStep.PreviousJobs, 'acme corp'));
        actualList.push(getResultByCurrentStep(CvExecutorStep.PreviousJobsTitle, 'Front-end developer'));
        actualList.push(getResultByCurrentStep(CvExecutorStep.PreviousJobsYears, '2'));
        actualList.push(getResultByCurrentStep(CvExecutorStep.PreviousJobsConfirm, ''));

        actualList.push(getResultByCurrentStep(CvExecutorStep.PreviousJobs, 'example studios'));
        actualList.push(getResultByCurrentStep(CvExecutorStep.PreviousJobsTitle, 'Junior developer'));
        actualList.push(getResultByCurrentStep(CvExecutorStep.PreviousJobsYears, '1'));
        actualList.push(getResultByCurrentStep(CvExecutorStep.PreviousJobsConfirm, 'no'));

        expect(actualList).toMatchSnapshot();
    });
});
