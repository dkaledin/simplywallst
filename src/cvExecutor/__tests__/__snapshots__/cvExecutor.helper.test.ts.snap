// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getResultByCurrentStep should return expected response for every receive command 1`] = `
Array [
  Object {
    "message": Array [
      "Welcome to the CV terminal command!",
    ],
    "next": Object {
      "prompt": "job_title: (Front-end Developer)",
      "step": 1,
    },
  },
  Object {
    "next": Object {
      "prompt": "given_name:",
      "step": 2,
    },
    "result": Object {
      "job_title": "Front-end Developer",
    },
  },
  Object {
    "next": Object {
      "prompt": "family_name:",
      "step": 3,
    },
    "result": Object {
      "given_name": "Warren",
    },
  },
  Object {
    "next": Object {
      "prompt": "email:",
      "step": 4,
    },
    "result": Object {
      "family_name": "Buffett",
    },
  },
  Object {
    "next": Object {
      "prompt": "mobile:",
      "step": 5,
    },
    "result": Object {
      "email": "tupaclives1985@gmail.com",
    },
  },
  Object {
    "next": Object {
      "prompt": "portfolio_url:",
      "step": 6,
    },
    "result": Object {
      "mobile": "04 0404 4044",
    },
  },
  Object {
    "next": Object {
      "prompt": "bio:",
      "step": 7,
    },
    "result": Object {
      "portfolio_url": "https://warrenbuffett.id.au",
    },
  },
  Object {
    "next": Object {
      "prompt": "skills:",
      "step": 8,
    },
    "result": Object {
      "bio": "Hi I like building cool things...",
    },
  },
  Object {
    "next": Object {
      "prompt": "years:",
      "step": 9,
    },
    "result": Object {
      "skills": Object {
        "php": Object {},
      },
    },
  },
  Object {
    "next": Object {
      "prompt": "confidence_rating:",
      "step": 10,
    },
    "result": Object {
      "skills": Object {
        "php": Object {
          "years": 2,
        },
      },
    },
  },
  Object {
    "next": Object {
      "prompt": "technology:",
      "step": 11,
    },
    "result": Object {
      "skills": Object {
        "php": Object {
          "confidence_rating": 50,
        },
      },
    },
  },
  Object {
    "next": Object {
      "prompt": "Do You have any more skills? (yes)",
      "step": 12,
    },
    "result": Object {
      "skills": Object {
        "php": Object {
          "technology": Array [
            "symfony",
            "laravel",
          ],
        },
      },
    },
  },
  Object {
    "next": Object {
      "prompt": "skills:",
      "step": 8,
    },
  },
  Object {
    "next": Object {
      "prompt": "years:",
      "step": 9,
    },
    "result": Object {
      "skills": Object {
        "js": Object {},
      },
    },
  },
  Object {
    "next": Object {
      "prompt": "confidence_rating:",
      "step": 10,
    },
    "result": Object {
      "skills": Object {
        "js": Object {
          "years": 3,
        },
      },
    },
  },
  Object {
    "next": Object {
      "prompt": "technology:",
      "step": 11,
    },
    "result": Object {
      "skills": Object {
        "js": Object {
          "confidence_rating": 99,
        },
      },
    },
  },
  Object {
    "next": Object {
      "prompt": "Do You have any more skills? (yes)",
      "step": 12,
    },
    "result": Object {
      "skills": Object {
        "js": Object {
          "technology": Array [
            "typescript",
            "react",
            "redux",
            "node",
          ],
        },
      },
    },
  },
  Object {
    "message": Array [
      "Great!",
    ],
    "next": Object {
      "prompt": "previous_jobs:",
      "step": 13,
    },
  },
  Object {
    "next": Object {
      "prompt": "job_title:",
      "step": 14,
    },
    "result": Object {
      "previous_jobs": Object {
        "acme corp": Object {},
      },
    },
  },
  Object {
    "next": Object {
      "prompt": "years:",
      "step": 15,
    },
    "result": Object {
      "previous_jobs": Object {
        "acme corp": Object {
          "job_title": "Front-end developer",
        },
      },
    },
  },
  Object {
    "next": Object {
      "prompt": "Do You have any more previous jobs? (yes)",
      "step": 16,
    },
    "result": Object {
      "previous_jobs": Object {
        "acme corp": Object {
          "years": 2,
        },
      },
    },
  },
  Object {
    "next": Object {
      "prompt": "previous_jobs:",
      "step": 13,
    },
  },
  Object {
    "next": Object {
      "prompt": "job_title:",
      "step": 14,
    },
    "result": Object {
      "previous_jobs": Object {
        "example studios": Object {},
      },
    },
  },
  Object {
    "next": Object {
      "prompt": "years:",
      "step": 15,
    },
    "result": Object {
      "previous_jobs": Object {
        "example studios": Object {
          "job_title": "Junior developer",
        },
      },
    },
  },
  Object {
    "next": Object {
      "prompt": "Do You have any more previous jobs? (yes)",
      "step": 16,
    },
    "result": Object {
      "previous_jobs": Object {
        "example studios": Object {
          "years": 1,
        },
      },
    },
  },
  Object {
    "message": Array [
      "Json is ready to send!",
      "Please check it.",
    ],
    "next": Object {
      "prompt": "Is this Ok? (yes)",
      "step": 17,
    },
  },
]
`;
